src/frontend/src/optimizer/plan_node/stream_cdc_table_scan.rs

```diff
@@ -182,29 +182,102 @@ impl StreamCdcTableScan {
-            output_indices,
-            // The table desc used by backfill executor
-            state_table: Some(catalog),
-            cdc_table_desc: Some(self.core.cdc_table_desc.to_protobuf()),
-        });
 
-        PbStreamNode {
-            fields: self.schema().to_prost(),
+        // split the table name from the qualified table name, e.g. `database_name.table_name`
+        let (_, cdc_table_name) = self
+            .core
+            .cdc_table_desc
+            .external_table_name
+            .split_once('.')
+            .unwrap();
+
+        // jsonb filter expr: payload->'source'->>'table' = <cdc_table_name>
+        let filter_expr: ExprImpl = FunctionCall::new(
+            ExprType::Equal,
+            vec![
+                FunctionCall::new(
+                    ExprType::JsonbAccessStr,
+                    vec![
+                        FunctionCall::new(
+                            ExprType::JsonbAccess,
+                            vec![
+                                InputRef::new(0, DataType::Jsonb).into(),
+                                ExprImpl::literal_varchar("source".into()),
+                            ],
+                        )
+                        .unwrap()
+                        .into(),
+                        ExprImpl::literal_varchar("table".into()),
+                    ],
+                )
+                .unwrap()
+                .into(),
+                ExprImpl::literal_varchar(cdc_table_name.into()),
+            ],
+        )
+        .unwrap()
+        .into();
+
+        let filter_operator_id = self.core.ctx.next_plan_node_id();
+        let filter_stream_node = StreamNode {
+            operator_id: filter_operator_id.0 as _,
             input: vec![
                 // The merge node body will be filled by the `ActorBuilder` on the meta service.
                 PbStreamNode {
-                    node_body: Some(PbNodeBody::Merge(Default::default())),
+                    node_body: Some(PbNodeBody::Merge(MergeNode {
+                        upstream_fragment_id: upstream_source_id,
+                        ..Default::default()
+                    })),
                     identity: "Upstream".into(),
                     fields: upstream_schema.clone(),
                     stream_key: vec![], // not used
                     ..Default::default()
                 },
             ],
+            stream_key: vec![], // not used
+            append_only: true,
+            identity: "StreamCdcFilter".to_string(),
+            fields: upstream_schema.clone(),
+            node_body: Some(PbNodeBody::CdcFilter(CdcFilterNode {
+                search_condition: Some(filter_expr.to_expr_proto()),
+                upstream_source_id,
+                upstream_column_ids: upstream_column_ids.clone(),
+            })),
+        };
+
+        let exchange_operator_id = self.core.ctx.next_plan_node_id();
+        // Add a simple exchange node between filter and stream scan
```

Comment:
I guess it only enforces `Backfill` to be singleton. `Backfill -> Mview` is a hash shuffle since the data can be materialized as a distributed table.

Reply:
The motivation is to decouple from the distribution of `CdcFilter`. Since `CdcFilter` is no shuffle, `Backfill` will also be no shuffle if we don't insert a exchange here. IIUC, then all Backfill created by different Table job will be scheduled onto the same CN, which is not expected.